rp2: Add support for USER_C_MODULES into the cmake build system

Gadgetoid opened the pull request:
I've been preserving these files in our own fork of MicroPython since Graham wrote them, source - https://github.com/raspberrypi/micropython/commit/a8b3c0c26699562ffbe1f901c8e5abed0a95a258

This is the USER_C_MODULES support that our continuous-integration and deployment relies upon to bake our libraries (https://github.com/pimoroni/pimoroni-pico/tree/main/micropython/modules) into MicroPython for our own .uf2 release. It's been pretty robust up to this point, and I'd love to get it merged so I can target upstream MicroPython for releases rather than maintaining our own CI fork.

Gadgetoid committed:
ports/rp2/mpthreadoprt.h: Cast core_state to _mp_state_thread_t.

Required for user C++ code to build successfully against ports/rp2.

Signed-off-by: Phil Howard <phil@pimoroni.com>

Gadgetoid committed:
ports/rp2: Add support for USER_C_MODULES to CMake build system.

py/usermod.cmake:

* Creates a usermod target to hang user C/CXX modules from.
* Gathers sources from user C/CXX modules and libs for QSTR scan.

ports/rp2/CMakeLists.txt:

* Includes py/usermod.cmake.
* Links the resulting usermod library to the MicroPython target.

py/mkrules.cmake:

Add cxxflags to qstr.i.last custom command for CXX modules:

* MICROPY_CPP_FLAGS so CXX modules will find includes.
* -DNO_QSTR to fix fatal error missing "genhdr/qstrdefs.generated.h".

Usage:

The rp2 port can be linked against user C modules by running:

make USER_C_MODULES=/path/to/module/micropython.cmake

CMake will print a list of included modules.

Co-authored-by: Graham Sanderson <graham.sanderson@raspberrypi.org>
Co-authored-by: Michael O'Cleirigh <michael.ocleirigh@rivulet.ca>
Signed-off-by: Phil Howard <phil@pimoroni.com>

Gadgetoid committed:
examples/usercmodule: Add micropython.cmake to the C and CPP examples.

examples/usercmodule/micropython.cmake:

Root micropython.cmake file is responsible for including modules.

examples/usercmodule/cexample/micropython.cmake:
examples/usercmodule/cppexample/micropython.cmake:

Module micropython.cmake files define the target and link it to usermod.

Signed-off-by: Phil Howard <phil@pimoroni.com>

stinos commented:
Could you update the documentation (cmodules.rst) as well? Assuming that is needed?

Gadgetoid commented:
Yes, the approach for setting up a `usermod.cmake` differs quite substantially from `micropython.mk` so it should be documented. Working on it.

So far:

```patch
diff --git a/docs/develop/cmodules.rst b/docs/develop/cmodules.rst
index 2db1f65f2..db121f8f5 100644
--- a/docs/develop/cmodules.rst
+++ b/docs/develop/cmodules.rst
@@ -39,12 +39,16 @@ A MicroPython user C module is a directory with the following files:
   to find similar modules within the MicroPython tree and use them as examples.
 
 * ``micropython.mk`` contains the Makefile fragment for this module.
+* ``usermod.cmake`` contains the CMake configuration for this module.
 
   ``$(USERMOD_DIR)`` is available in ``micropython.mk`` as the path to your
   module directory. As it's redefined for each c module, is should be expanded
   in your ``micropython.mk`` to a local make variable,
   eg ``EXAMPLE_MOD_DIR := $(USERMOD_DIR)``
 
+  In ``usermod.cmake``, you may use ``${CMAKE_CURRENT_LIST_DIR}`` as the path to
+  the current module.
+
   Your ``micropython.mk`` must add your modules source files relative to your
   expanded copy of ``$(USERMOD_DIR)`` to ``SRC_USERMOD``, eg
   ``SRC_USERMOD += $(EXAMPLE_MOD_DIR)/example.c``
@@ -53,6 +57,25 @@ A MicroPython user C module is a directory with the following files:
   for header files), these should be added to ``CFLAGS_USERMOD`` for C code
   and to ``CXXFLAGS_USERMOD`` for C++ code.
 
+  Your ``usermod.cmake`` should define an ``INTERFACE`` style library and
+  associate your source files, compile definitions and include directories
+  with it. The library should then be linked to ``usermod``.
+
+  .. code-block:: cmake
+
+      add_library(usermod_cexample INTERFACE)
+
+      target_sources(usermod_cexample INTERFACE
+          ${CMAKE_CURRENT_LIST_DIR}/examplemodule.c
+      )
+
+      target_include_directories(usermod_cexample INTERFACE
+          ${CMAKE_CURRENT_LIST_DIR}
+      )
+
+      target_link_libraries(usermod INTERFACE usermod_cexample)
+
+
   See below for full usage example.
 
 
@@ -70,7 +93,9 @@ and has a source file and a Makefile fragment with content as descibed above::
        └──usercmodule/
           └──cexample/
              ├── examplemodule.c
-             └── micropython.mk
+             ├── micropython.mk
+             └── usermod.cmake
+
 
 Refer to the comments in these 2 files for additional explanation.
 Next to the ``cexample`` module there's also ``cppexample`` which
@@ -97,10 +122,13 @@ applying 2 modifications:
       ├── modules/
       │   └──example1/
       │       ├──example1.c
-      │       └──micropython.mk
+      │       ├──micropython.mk
+      │       └──usermod.cmake
       │   └──example2/
       │       ├──example2.c
-      │       └──micropython.mk
+      │       ├──micropython.mk
+      │       └──usermod.cmake
+      │   └──usermod.cmake
       └── micropython/
           ├──ports/
          ... ├──stm32/
@@ -109,8 +137,10 @@ applying 2 modifications:
 
   with ``USER_C_MODULES`` set to the ``my_project/modules`` directory.
 
+  A top level ``usermod.cmake`` should ``include`` all of your modules.
+
 - all modules found in this directory will be compiled, but only those
-  which are explicitly enabled will be availabe for importing. Enabling a
+  which are explicitly enabled will be available for importing. Enabling a
   module is done by setting the preprocessor define from its module
   registration to 1. For example if the source code defines the module with


```

Gadgetoid committed:
docs/develop/cmodules.rst: Document micropython.cmake.

Documents the micropython.cmake file required to make user C modules
compatible with the CMake build system.

Signed-off-by: Phil Howard <phil@pimoroni.com>

stinos commented on docs/develop/cmodules.rst:
```diff
@@ -70,9 +94,11 @@ and has a source file and a Makefile fragment with content as descibed above::
+
 
-Refer to the comments in these 2 files for additional explanation.
+Refer to the comments in these 3 files for additional explanation.
```
Maybe just remove the number completely, more maintainable.

Gadgetoid replied:
Agreed!

stinos commented on docs/develop/cmodules.rst:
```diff
@@ -53,6 +53,30 @@ A MicroPython user C module is a directory with the following files:
+  In ``usermod.cmake``, you may use ``${CMAKE_CURRENT_LIST_DIR}`` as the path to
+  the current module.
+
+  Your ``usermod.cmake`` should define an ``INTERFACE`` style library and
```
What is an 'interface style library' ? (web search returns no sensible results)

gadirotenberg replied:
https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#interface-libraries

gadirotenberg replied:
> What is an 'interface style library' ? (web search returns no sensible results)

https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#interface-libraries

stinos replied:
Ok, perhaps just calling it "interface library" would be clearer then.

Gadgetoid replied:
I've made this change, in addition to retrospectively adding some comments to the .cmake files that - hopefully - clarify this. (And since it's mentioned the files include comments... they probably should include comments)

stinos replied:
> they probably should include comments

Hehe fair indeed.

stinos commented on examples/usercmodule/cexample/usermod.cmake:
```diff
@@ -0,0 +1,20 @@
+)
+
+# Link our INTERFACE library to the usermod target.
+target_link_libraries(usermod INTERFACE usermod_cexample)
```
This and other files miss a newline et the end. Not really an issue but for consistency with the rest of the codebase probably best to add it.

stinos commented:
Do you think it wuold be possible to reuse micropy_usermod.cmake for other ports? In which case it might go in the py/ directory (or maybe tools/, not usre)

Gadgetoid commented:
Looks like the pre-existing CMake tooling targets CMake >=3.5 while the rp2 targets CMake >=3.12. Guessing it was easier to copy the relevant files and change them to suit rp2 than update the canonical files and bump the required CMake version (I dread to think what knock-on effects that might have?). I suppose I can point the esp32 tooling at this file and see what happens, but I'm in a little over my head here now.

While maintaining separate ways of doing the same thing feels bad, it looks like the path of least resistance in this case.

Gadgetoid commented:
Okay I tried to build the esp32 port using this user C modules approach and it's totally incompatible. Something to do with the ESP idf's use of CMake or re-implementation of CMake... I'm unsure. If I was in over my head before, I've clipped through the ground and fallen into the void now. The long and short of it is that `add_library` results in the error "add_library command is not scriptable" and since it's fairly fundamental to this approach - and ESP idf-specific workarounds would defeat the point of it being placed in `py/` - I don't think there's anywhere I can go from here. I'd welcome insight from @mocleiri though.

stinos commented:
No big deal, was just wondering whether this would be possible at all.

Gadgetoid commented:
Yeah it'd be nice to have a canonical, shared way of handling things in CMake and it was a good idea. Melted my brain, though :laughing:  - thanks for your feedback here, by the way!

mocleiri commented:
@stinos @Gadgetoid I was able to take this approach and apply it to the esp32 firmware build.

I got the same error about "add_library command is not scriptable" but was able to resolve it by wrapping the logic in micropy_usermod.cmake with this:

if(NOT CMAKE_BUILD_EARLY_EXPANSION)
endif()

see: https://github.com/espressif/esp-idf/issues/4540#issuecomment-568504410

I modified the ports/esp32/main/CMakeLists.txt file in the same locations as the rp2 file and it seems to have worked.
![image](https://user-images.githubusercontent.com/250942/109255705-8f82b200-77c2-11eb-9968-da0ac7a5f3f2.png)

Gadgetoid commented:
Awesome! Thank you.

I guess I need to move `ports/rp2/micropy_usermod.cmake` to `py/usermod.cmake`.

Gadgetoid commented:
@mocleiri does the following still work for you. I've added some output to more closely match the docs:

```cmake
# Create a target for all user modules to link against.
add_library(usermod INTERFACE)

function(usermod_gather_sources VARNAME INCLUDED_VARNAME LIB)
    if (NOT ${LIB} IN_LIST ${INCLUDED_VARNAME})
        list(APPEND ${INCLUDED_VARNAME} ${LIB})
        set(${INCLUDED_VARNAME} ${${INCLUDED_VARNAME}} PARENT_SCOPE)
    endif()
    get_target_property(new_sources ${LIB} INTERFACE_SOURCES)
    if (new_sources)
        list(APPEND ${VARNAME} ${new_sources})
    endif()
    get_target_property(new_sources ${LIB} INTERFACE_LIBRARIES)
    get_target_property(trans_depend ${LIB} INTERFACE_LINK_LIBRARIES)
    if (trans_depend)
        foreach(SUB_LIB ${trans_depend})
            usermod_gather_sources(${VARNAME} ${INCLUDED_VARNAME} ${SUB_LIB})
        endforeach()
    endif()
    set(${VARNAME} ${${VARNAME}} PARENT_SCOPE)
    set(${INCLUDED_VARNAME} ${${INCLUDED_VARNAME}} PARENT_SCOPE)
endfunction()

# Include CMake files for user modules.
if (USER_C_MODULES)
    foreach(USER_C_MODULE_PATH ${USER_C_MODULES})
        message("Including User C Module(s) from ${USER_C_MODULE_PATH}/usermod.cmake")
        include(${USER_C_MODULE_PATH}/usermod.cmake)
    endforeach()
endif()

# Recursively gather sources for QSTR scanning - doesn't support generator expressions.
usermod_gather_sources(SOURCE_USERMOD found_mobules usermod)

# Report found modules.
list(REMOVE_ITEM found_mobules "usermod")
list(JOIN found_mobules ", " found_mobules)
message("Found User C Module(s): ${found_mobules}")
```

Gadgetoid commented:
Note: I'm still rewriting history and force-pushing to keep the commits clean and concise. Don't rely too heavily on my branch.

Example output from targeting our work-in-progress drivers:

```
Including User C Module(s) from /home/phil/Development/pico/pimoroni-pico/micropython/modules//usermod.cmake
PIOASM will need to be built
Found User C Module(s): usermod_breakout_3dof, usermod_breakout_6dof, usermod_breakout_9dof, usermod_breakout_ads1015, usermod_breakout_as7262, usermod_breakout_bh1745, usermod_breakout_bme280, usermod_breakout_bme680, usermod_breakout_bmp280, usermod_breakout_colourlcd160x80, usermod_breakout_colourlcd240x240, usermod_breakout_dotmatrix, usermod_breakout_drv8830, usermod_breakout_gps, usermod_breakout_haptic, usermod_breakout_ioexpander, usermod_breakout_ltr559, usermod_breakout_matrix11x7, usermod_breakout_max30105, usermod_breakout_mcp9600, usermod_breakout_mlx90640, usermod_breakout_oled128x128i2c, usermod_breakout_oled128x128spi, usermod_breakout_pmw3901, usermod_breakout_rgbencoder, usermod_breakout_rgbmatrix5x5, usermod_breakout_rgbpotentiometer, usermod_breakout_rtc, usermod_breakout_sgp30, usermod_breakout_switch, usermod_breakout_tof, usermod_breakout_trackball, usermod_breakout_veml6075, usermod_pico_display, usermod_pico_explorer, usermod_pico_rgb_keypad, usermod_pico_scroll, usermod_pico_unicorn
```

There's no requirement to prefix libraries with `usermod_`, but it makes it easier for us to avoid CMake target name collisions.

Gadgetoid commented:
Latest changes to documentation + blank line endings for usermod.cmake files: https://github.com/micropython/micropython/compare/214be3818e103de649a7b552c6d75e1129b20288..b4404bcb3c7907acc3b1ff2cf6f0ec9135d49501

Latest changes to `py/usermod.cmake` (and, of course I've made a silly typo) - https://github.com/micropython/micropython/compare/148da3a35902a11378c4b3cffec2540ffe111545..214be3818e103de649a7b552c6d75e1129b20288

mocleiri commented:
@Gadgetoid 

Yes, I'll need ports/rp2/micropy_usermod.cmake moved to py/usermod.cmake.

I'll pick up whatever the latest version is from your branch for now (when I do it this evening).  Then update my pull request to be from master with the py/usermod.cmake file and changes to the esp32 cmake files.

This should give you the freedom to rewrite the branch as necessary.

mocleiri commented:
My first test above only used the c example.  I just expanded to use the cpp example and it breaks due to include path issues.  The reason is related to the cflags passed to the qstr generation steps.  

The esp32 port is using py/mkrules.cmake but the rp2 port has its own copy in ports/rp2/micropy_rules.cmake.  In the esp32 port the include path for the user modules is not getting included in the cflags passed to the qstr generator which then breaks the compile when the cpp example sources can't resolve their header files.

In the esp32 port the command is:
https://github.com/micropython/micropython/blob/master/py/mkrules.cmake#L72

In the rp2 port the command is:
https://github.com/micropython/micropython/blob/master/ports/rp2/micropy_rules.cmake#L42

In the esp32 port I need to add a way to pass up the include flags and declaration flags from the module upward so that they can be appended to MICROPY_CPP_FLAGS cmake variable.

The esp32 makefile already has such accumulation logic for processing the idf components used it just needs to be extended or to use the existing MICROPY_CPP_INC_EXTRA and MICROPY_CPP_DEF_EXTRA from the usermodules. See: https://github.com/micropython/micropython/blob/master/ports/esp32/main/CMakeLists.txt#L169

It looks like esp32 and zephyr ports use py/mkrules.cmake and zeyphr adds directly to MICROPY_CPP_FLAGS_EXTRA here:
https://github.com/micropython/micropython/blob/master/ports/zephyr/CMakeLists.txt#L83

In the rp2 port it uses the cmake C_FLAGS variable but I don't see it being set anywhere.  I'm going to try this variable in py/mkrules.cmake to see if it works and if not then bubble up the include paths from the usermodules.

Gadgetoid commented:
The recursive function in `usermod.cmake` could be, I think, trivially expanded to collect all the CFLAGS/CPPFLAGS into another variable alongside `SOURCE_USERMOD` since it's already using `get_target_property` on every included library and dependency.

Perhaps these should be `CLAGS_USERMOD` and `CPPFLAGS_USERMOD`.

I'm a little concerned that the rp2 port doesn't include these, since it would suggest qstrs might be generated from a different interpretation of the pre-processor macros than the actual build itself. I'm not familiar enough with the MicroPython build and the concept of qstrs to know if this is a huge problem though.

mocleiri commented:
@Gadgetoid as you suggested I modified the recursive function to also collect the user module include directories: https://github.com/micropython/micropython/pull/6956/commits/de80a488b5b945cc7b67671655bb0f2119522cfa#diff-82357b91a9f2b8785acebd96061b5cf0be2a03036abfe7f41d5a2a1fbe46fe8e

I was able to get the c++ example to work by taking the rp2 command used to make the qstr.i.last file:
https://github.com/micropython/micropython/pull/6956/commits/f740c036ee96ee7b3c37771d552d5a208e6ba6cd#diff-00cee5de6f52e817db60265799e2973a0de5e111521b807d7b8bc7b355132569

For this last change I think the zephyr port would need to test to see that their firmware still builds because they use the **py/mkrules.cmake** file with esp32.

Gadgetoid commented:
I guess the way to proceed now is for me to test the above reference modifications don't totally explode the rp2 setup, hope this PR gets merged and for you to rebase #6956 on top of the result, including the changes to `py/usermod.cmake`?

Though I'm not opposed to sweeping the changes from https://github.com/micropython/micropython/commit/de80a488b5b945cc7b67671655bb0f2119522cfa#diff-82357b91a9f2b8785acebd96061b5cf0be2a03036abfe7f41d5a2a1fbe46fe8e into my copy of `py/usermod.cmake` and updating the co-authorship on the commit.

I tend to favour the latter, since I think your includes changes being merged with the creation of `py/usermod.cmake` will discourage anyone from rolling their own approach. I'm probably tremendously overthinking this.

The TLDR: Is there anything else I need to do?

@jsimonrichard since your issue #7006 relates to this, if you have an opportunity to try these changes I'd love to know if they work for you.

jsimonrichard commented:
@Gadgetoid I cloned https://github.com/pimoroni/micropython/tree/continuous-integration and tried to build the rp2 firmware (without defining USER_C_MODULES), but I got the following error.

```
/home/jsimonrichard/Desktop/Micro-Code/pimoroni-micropython/ports/rp2/main.c:37:10: fatal error: tusb.h: No such file or directory
   37 | #include "tusb.h"
      |          ^~~~~~~~
```
I've already updated the submodules using `git submodule update lib/tinyusb` and `git submodule update lib/pico-sdk` in the root directory of the project.

Is there something else I need to install?

Gadgetoid commented:
@jsimonrichard IIRC you *might* have to update/init the Tiny USB submodule *inside* `pico-sdk`. ie: `lib/pico-sdk/lib/tinyusb`.

When we actually run CI against this branch, it does:

https://github.com/pimoroni/pimoroni-pico/blob/292e2b52afce11daccbe29897d716b06f38c89e6/.github/workflows/micropython.yml#L52-L55

IE: descend into the `pico-sdk` submodule and run `git submodule update --init` in there too.

You may also have to delete your `port/rp2/build` directory manually sometimes. It'll occasionally get into a pickle.

jsimonrichard commented:
@Gadgetoid
Alright, I was able to build the firmware without adding any extra modules. After that, I tried to build it with the ulab module using this command:
```
make USER_C_MODULES=../../../ulab/ CFLAGS=-DMODULE_ULAB_ENABLED all
```
It compiled, but `Included User C Module from ...` is still missing from stdout. I also flashed the firmware onto a pico, but `import ulab` didn't work.

This is the directory structure I'm working with:
```
ulab/
    code/
        micropython.mk
        ulab.c
        
pimoroni-micropython/
    ports/
         rp2/
            makefiles, c code, etc.
```

Gadgetoid commented:
See https://github.com/micropython/micropython/pull/6960/commits/7e968d5819160889fc08c27e0ac847ae4af41e14 for some examples of the required `usermod.cmake` file, that is used in lieu of `micropython.mk` for CMake builds.

And, to save you some small headache, here's what one might look like for ulab:

```cmake
add_library(usermod_ulab INTERFACE)

target_sources(usermod_ulab INTERFACE
    ${CMAKE_CURRENT_LIST_DIR}/scipy/optimize/optimize.c
    ${CMAKE_CURRENT_LIST_DIR}/scipy/signal/signal.c
    ${CMAKE_CURRENT_LIST_DIR}/scipy/special/special.c
    ${CMAKE_CURRENT_LIST_DIR}/ndarray_operators.c
    ${CMAKE_CURRENT_LIST_DIR}/ulab_tools.c
    ${CMAKE_CURRENT_LIST_DIR}/ndarray.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/approx/approx.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/compare/compare.c
    ${CMAKE_CURRENT_LIST_DIR}/ulab_create.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/fft/fft.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/fft/fft_tools.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/filter/filter.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/linalg/linalg.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/linalg/linalg_tools.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/numerical/numerical.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/poly/poly.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/stats/stats.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/transform/transform.c
    ${CMAKE_CURRENT_LIST_DIR}/numpy/vector/vector.c

    ${CMAKE_CURRENT_LIST_DIR}/numpy/numpy.c
    ${CMAKE_CURRENT_LIST_DIR}/scipy/scipy.c
    ${CMAKE_CURRENT_LIST_DIR}/user/user.c
    ${CMAKE_CURRENT_LIST_DIR}/utils/utils.c
    ${CMAKE_CURRENT_LIST_DIR}/ulab.c
)

target_include_directories(usermod_ulab INTERFACE
    ${CMAKE_CURRENT_LIST_DIR}
)

target_compile_definitions(usermod_ulab INTERFACE
    -DMODULE_ULAB_ENABLED=1
)

target_link_libraries(usermod INTERFACE usermod_ulab)

```

Then your make command would be `make USER_C_MODULES=../../../ulab/code` - you don't need to explicitly enable the module, since `ulab` does this for you (at least it looks like it does) and so I've ported this behaviour to the CMake setup.

jsimonrichard commented:
I copied that file into `ulab/code/` as `usermod.cmake`, but it's still not working. Did you get a chance to test it yourself?

Gadgetoid commented:
Yes, I have a version built with ulab and our house modules:

```python
>>> import ulab
>>> dir(ulab)
['__class__', '__name__', '__version__', 'dtype', 'numpy', 'scipy', 'utils']
```

And am potentially shipping a change to our custom .uf2 build system to produce a build with `ulab` included (though it's pretty huge in relative terms!):

https://github.com/pimoroni/pimoroni-pico/pull/88

We should probably switch this conversation over to https://github.com/v923z/micropython-ulab/pull/346 since it's more about supporting `ulab` with this system than this particular change. If you post the steps you followed and the output there I can help guide you- and hopefully by extensions others- toward successfully building using this tooling.

v923z commented:
> And am potentially shipping a change to our custom .uf2 build system to produce a build with `ulab` included (though it's pretty huge in relative terms!):

If you don't need all functions, you can configure your build here: https://github.com/v923z/micropython-ulab/blob/master/code/ulab.h

v923z commented:
@stinos @Gadgetoid Is `micropython` going to switch to `cmake` on all ports, or only on `ESP32`, and `rp2`? In the latter case, would it be possible to have a single source of truth, i.e., generate both the `micropython.mk` make fragment, and the `usermod.cmake` file from a single list of files?

stinos commented:
@v923z not my call, but I highly doubt all ports are going to switch, at least not for the time being: would be quite a lot of work, with not that much benefit

v923z commented:
@stinos But then the approach in this PR should be revised. Ports should have a single file for the source file list. Having separate make fragments, and cmake fragments is going to bite someone in the future.

Gadgetoid commented:
@v923z the whole concept of having Make and CMake builds in MicroPython has always struck me as something of a pain, but what's done is done and I believe `RP2` and `ESP32` both lean heavily on their respective CMake SDKs to make some of their features available to C/C++ modules.

I'm not aware (albeit my CMake and Make knowledge isn't very deep) of an easy way to make a single source of truth like you describe, but I can certainly see the benefit. The ideal solution would be for CMake to support `micropython.mk` files and recursive scanning of the given directories, just like Make, but that may be a pipedream and I only have so much time to invest in this PR.

This looks interesting - https://gist.github.com/tusharpm/d71dd6cab8a00320ddb48cc82bf7f64c - but it would still require the creation of a CMake target with linked dependencies. It might be possible to make a "drop in" `usermod.cmake` that will work in many cases, but I only see that as more bitey, not less.

On the flip side, *requiring* the addition of a `usermod.cmake` does mean a particular module author has - presumably - vetted their module for compatibility with the CMake approach (I'm not sure what could be *incompatible* but this stuff always surprises me).

I guess a python "turn this micropython.mk into a usermod.cmake" tool might be useful, and could be included in the MicroPython repository?

I'm all ears for a workable alternative since I totally understand where you're coming from, but right now I'm just playing the cards I've been dealt. (and trying to upstream code I'm relying on so I don't have to maintain a fork of MicroPython and produce out-of-date builds)

v923z commented:
@Gadgetoid 

> This looks interesting - https://gist.github.com/tusharpm/d71dd6cab8a00320ddb48cc82bf7f64c - but it would still require the creation of a CMake target with linked dependencies. It might be possible to make a "drop in" `usermod.cmake` that will work in many cases, but I only see that as more bitey, not less.

It seems to me that the structure of `usermod.cmake` is quite simple, and it doesn't contain much logic, at least, as far as user modules are concerned. 

> 
> On the flip side, _requiring_ the addition of a `usermod.cmake` does mean a particular module author has - presumably - vetted their module for compatibility with the CMake approach (I'm not sure what could be _incompatible_ but this stuff always surprises me).

But this doesn't solve the problem of maintainability: every time a new file is added to the source list, both `usermod.cmake`, and `micropython.mk` have to be updated. It would also require the developers of third-party libraries to continuously check changes related to the build system. The whole idea of `micropython` is that you insert an abstraction layer between the hardware and the user, so that the user doesn't have to care about the implementation details. Similarly, the compilation of firmware-only libraries should be independent of the hardware. 

> I guess a python "turn this micropython.mk into a usermod.cmake" tool might be useful, and could be included in the MicroPython repository?

I think this is how it should be done.

Gadgetoid commented:
@dpgeorge do you have any insight here? We'll be setting precedent, so I'd like to get it at least somewhat right.

@v923z  I can throw together a Python tool and it feels like a workable solution for the simpler cases, but upon reflection I'm worried that making it comprehensive will be next to impossible, a cursory check for pre-existing tools doesn't yield much of use.

I've collected some specimen `micropython.mk` files and they can get quite complex, for example: https://github.com/Kevingislason/abacus_wallet/blob/00723de2d702fb1d3297f4e91c95569a1fcc0250/usermods/udisplay_f469/micropython.mk

While CMake can "glob" for source files, it's broadly recommended against since it doesn't detect newly added files. *However* with the assumption that a user will not add files to a C module, this might work as a throw-in-and-forget for your case:

```cmake
add_library(usermod_ulab INTERFACE)

file(GLOB_RECURSE ULAB_SOURCES ${CMAKE_CURRENT_LIST_DIR}/*.c)

message(WARNING ${CMAKE_CURRENT_LIST_DIR})
message(WARNING ${ULAB_SOURCES})

target_sources(usermod_ulab INTERFACE
    ${ULAB_SOURCES}
)

target_include_directories(usermod_ulab INTERFACE
    ${CMAKE_CURRENT_LIST_DIR}
)

target_compile_definitions(usermod_ulab INTERFACE
    -DMODULE_ULAB_ENABLED=1
)

target_link_libraries(usermod INTERFACE usermod_ulab)
```

v923z commented:
> I've collected some specimen `micropython.mk` files and they can get quite complex, for example: https://github.com/Kevingislason/abacus_wallet/blob/00723de2d702fb1d3297f4e91c95569a1fcc0250/usermods/udisplay_f469/micropython.mk

Uh-hu, this is complicated... 

> 
> While CMake can "glob" for source files, it's broadly recommended against since it doesn't detect newly added files. _However_ with the assumption that a user will not add files to a C module, this might work as a throw-in-and-forget for your case:

Do you want to move this to https://github.com/v923z/micropython-ulab/pull/346?

Gadgetoid commented:
@mocleiri referring back to https://github.com/micropython/micropython/pull/6960#issuecomment-792761489 would you prefer I incorporate your `py/usermod.cmake` tweaks here, or leave them to your PR to build on top of this one?

@stinos @gadirotenberg is there anything else blocking a merge?

dpgeorge commented:
Thanks @Gadgetoid for this contribution, it looks really good.  I need to go through this discussion then will provide more feedback.

mocleiri commented:
> @mocleiri referring back to [#6960 (comment)](https://github.com/micropython/micropython/pull/6960#issuecomment-792761489) would you prefer I incorporate your `py/usermod.cmake` tweaks here, or leave them to your PR to build on top of this one?
> 

@Gadgetoid I'm fine if you want to fold those changes into this pull  request.

I need to rebase my pull request anyways so can do that after you be incorporate the changes.

Also feel free to adjust as you want.  I think I have some unnecessary whitespace and logging.

Gadgetoid commented:
Rebasing on top of https://github.com/micropython/micropython/commit/a9140ab09b7f2b878972a86dabc12d00ddaa463e has broken this beyond my ability to fix.

It now appears to need (though I don't understand how or why this suddenly differs) the same gather-include-directories fix required for ESP32, otherwise the build against our module tree will fail with:

```
In file included from /home/phil/Development/pico/pimoroni-pico/micropython/modules/ulab/code/numpy/filter/filter.c:23:
/home/phil/Development/pico/pimoroni-pico/micropython/modules/ulab/code/numpy/filter/../../scipy/signal/signal.h:16:10: fatal error: ulab.h: No such file or directory
   16 | #include "ulab.h"
      |          ^~~~~~~~
```

And if I gather include directories from the user modules and attach them to the build and fail with, suggesting that Pico SDK includes are now not being found for reasons I can't divine:

```
/home/phil/Development/pico/pimoroni-pico/micropython/modules/pico_scroll/pico_scroll.cpp:1:10: fatal error: hardware/spi.h: No such file or directory
    1 | #include "hardware/spi.h"
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
```

We're still relying upon our `continuous-integration` branch to produce custom .uf2 builds with C modules so I've temporarily pushed up to a new branch in case anyone has some ideas: https://github.com/pimoroni/micropython/commits/continuous-integration-broken

And here's the commit that adds `usermod.cmake` - https://github.com/pimoroni/micropython/commit/a23e1ecdb0ad45523cd7f4c97b391be23cd8470d

mocleiri commented:
@Gadgetoid the only difference I can see is that in ports/esp32/main/CMakeLists.txt (line 13)
In addition to using the accumulated include paths I also add $MICROPY_PY_DIR into MICROPY_CPP_INC_EXTRA:
https://github.com/micropython/micropython/pull/6956/commits/de80a488b5b945cc7b67671655bb0f2119522cfa#diff-82357b91a9f2b8785acebd96061b5cf0be2a03036abfe7f41d5a2a1fbe46fe8eR13

I don't see that set in ports/rp2/CMakeLists.txt although I didn't look at the other surrounding files.

Can you try appending $MICROPY_PY_DIR to MICROPY_CPP_INC_EXTRA to see if it makes a difference?

MICROPY_PY_DIR is the same as $MICROPY_DIR/py and is setup in py/py.cmake file.

Gadgetoid commented:
Thanks @mocleiri ... but no dice. This seems to be a problem unique to the RP2 port whereupon user modules use some of the RP2 or Pico SDK libraries. They used to resolve implicitly and built fine, but now the individual user module builds all fail to find various Pico SDK headers.

I seem to be having some weird build errors even with the code that's supposed to work- and *does* still work in our CI - so I might have some debugging to do.

v923z commented:
@Gadgetoid If this issue concerns only `ulab`, I am willing to add a temporary patch, so that you don't have to maintain a separate `micropython` fork. But I have the feeling that the problem is much more fundamental.

Gadgetoid commented:
@v923z yeah it appears to be much more fundamental and - in fact- `ulab` appears to be the one thing that *is* working, having reset my local build system back to a usable state. `ulab` builds were actually failing with the code that I expected to work, but working in CI which pointed a pretty clear finger at *something* on my machine being out of whack.

Sadly that *something* is not at all related to the Pico SDK issues I'm now seeing when trying to build our own modules.

No matter how I slice it, or where I try to stick include directories, or how much trickery I add/remove from the CMake build I can't get past this:

```
pimoroni-pico/micropython/modules/pico_scroll/pico_scroll.cpp:1:10: fatal error: hardware/spi.h: No such file or directory
    1 | #include "hardware/spi.h"
      |          ^~~~~~~~~~~~~~~~
```

Which has been introduced by the switch from custom `RP2`-only CMake files, to adopting the standard ones in `py/`.

All of the Pico SDK components are still linked to the MicroPython target as before with only minor indirection, and I can't see anything else that looks like it might be the culprit. Very odd.

v923z commented:
> @v923z yeah it appears to be much more fundamental and - in fact- `ulab` appears to be the one thing that _is_ working,

:grin:

Gadgetoid commented:
Wrangled some git-fu to give me an updated CI branch without the offending commit, which also allows me to get a succinct file difference between working (`continuous-integration-new`) and not-working (`continuous-interation-broken`) branches:

https://github.com/pimoroni/micropython/compare/continuous-integration-new..pimoroni:continuous-integration-broken

Looking closer, the issue is with the `Generating genhdr/qstr.i.last` step that very clearly *has* changed in the switch to using the standard CMake components.

In fact, knowing this and comparing the function with the previous version lets me do some experimentation and I believe I've got my setup at least getting somewhat *further* with this possibly contentious patch:

```patch
diff --git a/py/mkrules.cmake b/py/mkrules.cmake
index 9c4c4afab..f5daf1bef 100644
--- a/py/mkrules.cmake
+++ b/py/mkrules.cmake
@@ -69,7 +69,7 @@ add_custom_command(
 # it was last run, but it looks like it's not possible to specify that with cmake.
 add_custom_command(
     OUTPUT ${MICROPY_QSTRDEFS_LAST}
-    COMMAND ${Python3_EXECUTABLE} ${MICROPY_PY_DIR}/makeqstrdefs.py pp ${CMAKE_C_COMPILER} -E output ${MICROPY_GENHDR_DIR}/qstr.i.last cflags ${MICROPY_CPP_FLAGS} -DNO_QSTR sources ${MICROPY_SOURCE_QSTR}
+    COMMAND ${Python3_EXECUTABLE} ${MICROPY_PY_DIR}/makeqstrdefs.py pp ${CMAKE_C_COMPILER} -E \$\(C_INCLUDES\) \$\(C_FLAGS\) -DNO_QSTR output ${MICROPY_GENHDR_DIR}/qstr.i.last cflags ${MICROPY_CPP_FLAGS} -DNO_QSTR sources ${MICROPY_SOURCE_QSTR}
     DEPENDS ${MICROPY_MODULEDEFS}
         ${MICROPY_SOURCE_QSTR}
     VERBATIM
 ```
 
 Though the addition of `-Werror` very quickly rains on my parade.
 
@dpgeorge any idea why the addition of a *second* `-DNO_QSTR` might have been required in this case? Without it I'd get the build failing on dependencies that didn't exist yet (`genhdr/qstrdefs.generated.h`):

```
[  7%] Generating genhdr/qstr.i.last
In file included from /home/phil/Development/pico/micropython/py/obj.h:33,
                 from /home/phil/Development/pico/micropython/py/mpstate.h:35,
                 from /home/phil/Development/pico/micropython/py/runtime.h:29,
                 from /home/phil/Development/pico/fresh/pimoroni-pico/micropython/modules/pico_scroll/pico_scroll.h:2,
                 from /home/phil/Development/pico/fresh/pimoroni-pico/micropython/modules/pico_scroll/pico_scroll.cpp:13:
/home/phil/Development/pico/micropython/py/qstr.h:42:14: fatal error: genhdr/qstrdefs.generated.h: No such file or directory
   42 |     #include "genhdr/qstrdefs.generated.h"
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

The missing `C_INCLUDES` and `C_FLAGS` were what have been hosing my build and giving me a headache these past few days.


(guess we better fix those warnings)

Gadgetoid commented:
Hokay! My changes to the `qstr.i.last` generator have broken Zephr, I thought I was being smart getting it to work again but clearly I'm once again in over my head.

On the upside at least our CI branch and this PR is rebased on top of the latest changes and *mostly* working.

From my perspective- with my narrow focus upon getting our custom Pico SDK C++ modules building against rp2 - this is great. I can probably continue rebasing this from upstream to keep our CI and custom .uf2 releases from going *too* stale. Clearly there's something to work out here before this is merge-worthy.

kilograham commented:
I just saw @Gadgetoid mention this on Twitter - not sure where this currently stands 

> @v923z yeah it appears to be much more fundamental and - in fact- `ulab` appears to be the one thing that _is_ working, having reset my local build system back to a usable state. `ulab` builds were actually failing with the code that I expected to work, but working in CI which pointed a pretty clear finger at _something_ on my machine being out of whack.
> 
> Sadly that _something_ is not at all related to the Pico SDK issues I'm now seeing when trying to build our own modules.
> 
> No matter how I slice it, or where I try to stick include directories, or how much trickery I add/remove from the CMake build I can't get past this:
> 
> ```
> pimoroni-pico/micropython/modules/pico_scroll/pico_scroll.cpp:1:10: fatal error: hardware/spi.h: No such file or directory
>     1 | #include "hardware/spi.h"
>       |          ^~~~~~~~~~~~~~~~
> ```
> 
> Which has been introduced by the switch from custom `RP2`-only CMake files, to adopting the standard ones in `py/`.
> 
> All of the Pico SDK components are still linked to the MicroPython target as before with only minor indirection, and I can't see anything else that looks like it might be the culprit. Very odd.

do you have hardware_spi as a dependency in your usermod.cmake?

dpgeorge commented:
> The missing `C_INCLUDES` and `C_FLAGS` were what have been hosing my build and giving me a headache these past few days.

Unfortunately this won't work.  The idea with the core-provided cmake rules (that esp32 and zephyr ports use, and which I recently updated rp2 to use) is that cmake can be used with a different back end to make, ie ninja.  And these `C_INCLUDES`/`C_FLAGS` variables are unknown to ninja.

Changing the rp2 port to use the new core-provided cmake rules did really pull the rug out from under this PR, sorry about that!  I will see if I can fix things...

dpgeorge commented on py/mkrules.cmake:
```diff
@@ -69,7 +69,7 @@ add_custom_command(
 add_custom_command(
     OUTPUT ${MICROPY_QSTRDEFS_LAST}
-    COMMAND ${Python3_EXECUTABLE} ${MICROPY_PY_DIR}/makeqstrdefs.py pp ${CMAKE_C_COMPILER} -E output ${MICROPY_GENHDR_DIR}/qstr.i.last cflags ${MICROPY_CPP_FLAGS} -DNO_QSTR sources ${MICROPY_SOURCE_QSTR}
+    COMMAND ${Python3_EXECUTABLE} ${MICROPY_PY_DIR}/makeqstrdefs.py pp ${CMAKE_C_COMPILER} -E ${MICROPY_CPP_FLAGS} -DNO_QSTR output ${MICROPY_GENHDR_DIR}/qstr.i.last cflags ${MICROPY_CPP_FLAGS} -DNO_QSTR sources ${MICROPY_SOURCE_QSTR}
```
I think what's needed is to follow how `makeqstrdefs.py` is invoked in `py/mkrules.mk`, and add something like:
```
... cxxflags ${MICROPY_CPP_FLAGS} -DNO_QSTR ...
```
and if there are any additional C++ specific flags for the build, they also need to be added in that snippet.

Gadgetoid replied:
As need as I can tell, all the relevant include directories are added to `MICROPY_CPP_FLAGS` by your code at the bottom of `rp2`'s  `CMakeLists.txt` but unless I add `${MICROPY_CPP_FLAGS}` twice- following how the old `rp2` specific invocation of this command by adding it just after the `-E` flag.

From what I can understand these flags are being then passed *directly* to the C compiler command that `makeqsstrdefs.py` is invoking.

dpgeorge replied:
I tried moving the duplicated args (after the -E flag) to after a new `cxxflags` entry, and the rp2 port compiled cleanly with `make USER_C_MODULES=../../examples/usercmodule`

I guess your problem is with compiling ulab though?

Gadgetoid replied:
`ulab` somehow seems to work fine, compiling against our house C++ modules is where it falls flat since they leverage the Pico SDK.

I'll give this a go since I haven't tried `cxxflags` yet, I was totally missing the distinction between `cxxflags` and `cflags`.

dpgeorge replied:
I'm trying it now.  I seem to be able to build your entire repo of user-C-modules without issue, including ulab.

Gadgetoid replied:
Agreed, just tested it locally and force-pushed to see how Zephr reacts. Thank you!

dpgeorge replied:
But to build with `pimoroni-pico/micropython/modules` I did need to make some changes like this:
```diff
--- a/micropython/modules/pico_display/pico_display.cpp
+++ b/micropython/modules/pico_display/pico_display.cpp
@@ -2,7 +2,7 @@
 #include "hardware/sync.h"
 #include "pico/binary_info.h"
 
-#include "../../../pimoroni-pico/libraries/pico_display/pico_display.hpp"
+#include "../../../libraries/pico_display/pico_display.hpp"
 
 using namespace pimoroni;
 
```

Gadgetoid replied:
Not the case here, which is weird- it *just works*- since I've updated the `continuous-integration` branch I'll bump our automated build and see if it still works.

Gadgetoid replied:
Seems fine (assuming you can see this) - https://github.com/pimoroni/pimoroni-pico/runs/2174755007

Need to fix the `ulab` `usermod.cmake` though since I left some debug output in the PR :grimacing:

dpgeorge commented on docs/develop/cmodules.rst:
```diff
@@ -53,6 +53,30 @@ A MicroPython user C module is a directory with the following files:
   for header files), these should be added to ``CFLAGS_USERMOD`` for C code
   and to ``CXXFLAGS_USERMOD`` for C++ code.
 
+* ``usermod.cmake`` contains the CMake configuration for this module.
```
What was the reason to call it `usermod.cmake`?  Why not `micropython.cmake` to match `micropython.mk`?

Gadgetoid replied:
I think the reasoning is lost to the sands of time, since I'm carrying this changeset over from the now ghost Raspberry Pi repository. This has actually been bugging me for a while and you're right- `micropython.cmake` is a lot clearer. Only reason I haven't changed it yet is because I have to rename a whole bunch of files across three repositories.

dpgeorge replied:
I think it's worth changing now.  As you say, this PR sets a precedent.

dpgeorge commented on examples/usercmodule/cexample/usermod.cmake:
```diff
@@ -0,0 +1,21 @@
+
+# Enable the module automatically by adding the relevant compile definitions.
+target_compile_definitions(usermod_cexample INTERFACE
+    -DMODULE_CEXAMPLE_ENABLED=1
```
I don't think this needs the `-D` prefix.

Gadgetoid replied:
Today I learned. Will have to fix this up across the cmake snippets I've scattered to the four winds, too.

Gadgetoid replied:
Ran a build to verify this and - yep - you're right.

dpgeorge commented on py/usermod.cmake:
```diff
@@ -0,0 +1,52 @@
+if (USER_C_MODULES)
+    foreach(USER_C_MODULE_PATH ${USER_C_MODULES})
+        message("Including User C Module(s) from ${USER_C_MODULE_PATH}/usermod.cmake")
+        include(${USER_C_MODULE_PATH}/usermod.cmake)
```
How about `USER_C_MODULES` is a (list) of paths to `.cmake` fragments rather than a directory within which `usermod.cmake` is searched for?  So the include would just be `include(${USER_C_MODULE_PATH})`

Then the build would be:
```
make USER_C_MODULES=../../examples/usercmodule/usermod.cmake
```
This way there can be multiple cmake fragments in a directory and the user can select which one to use (or multiple ones as a list).

Gadgetoid replied:
Sensible!

dpgeorge commented on py/usermod.cmake:
```diff
@@ -0,0 +1,52 @@
+endif()
+
+# Recursively gather sources for QSTR scanning - doesn't support generator expressions.
+usermod_gather_sources(MICROPY_SOURCE_USERMOD MICROPY_INCLUDE_DIRECTORIES_USERMOD found_modules usermod)
```
How about renaming `MICROPY_INCLUDE_DIRECTORIES_USERMOD` to the shorter `MICROPY_INC_USERMOD`?  This would match the use of INC in, eg, `MICROPY_CPP_INC`.

Gadgetoid replied:
Agreed! It was something of a mouthful.

dpgeorge commented:
Since this feature looks a bit fragile, might be good to add a build of user-C-modules to the CI.  Eg:
```diff
--- a/tools/ci.sh
+++ b/tools/ci.sh
@@ -181,6 +181,8 @@ function ci_rp2_build {
     make ${MAKEOPTS} -C mpy-cross
     git submodule update --init lib/pico-sdk lib/tinyusb
     make ${MAKEOPTS} -C ports/rp2
+    make ${MAKEOPTS} -C ports/rp2 clean
+    make ${MAKEOPTS} -C ports/rp2 USER_C_MODULES=../../examples/usercmodule/usermod.cmake
 }

```

Gadgetoid committed:
tools/ci.sh: Add CI for CMake USER_C_MODULE support.

Builds the rp2 port against the example C and CXX modules.

Signed-off-by: Phil Howard <phil@pimoroni.com>

Gadgetoid commented:
Let's pretend we didn't see that :eyes:

Gadgetoid commented:
Okay I'm done making stupid mistakes. @mocleiri I've folded your code into `py/usermod.cmake` along with some other (hopefully minor) tweaks to prevent a case where it could recurse itself to oblivion. When you get a moment, can you confirm it still works for you? Thanks!

@dpgeorge I *think* I'm all up to date with your suggested changes. Note that the documentation still focusses on the "top level micropython.cmake that includes each module micropython.cmake" rather than specifying multiple modules on the command line. It could possibly be more explicit about the latter working, too.

dpgeorge commented on examples/usercmodule/cexample/micropython.cmake:
```diff
@@ -0,0 +1,21 @@
+# Enable the module automatically by adding the relevant compile definitions.
+target_compile_definitions(usermod_cexample INTERFACE
+    MODULE_CEXAMPLE_ENABLED=1
+)
```
What was the reasoning to enable the module automatically?  The makefile (.mk) way doesn't do this, so this will be an inconsistency.  Also, how would one disable (override) this value?

Gadgetoid replied:
I'm going to point to ulab here as probably the most high profile example to back up my reasoning:

https://github.com/v923z/micropython-ulab/blob/5890f0e193ef91ba584f114afc4fe3a9e8de6a99/code/micropython.mk#L33

But there's also:

https://github.com/devbis/st7789_mpy/blob/master/st7789/micropython.mk
https://github.com/mocleiri/tensorflow-micropython-examples/blob/5a8656b658017538166bfdd92c89f58124014a75/micropython-modules/audio_frontend/micropython.mk
https://github.com/glennrub/ubluepy/blob/3a1ae04c98a2a460aab84ba13294825bdd43596b/micropython.mk
https://github.com/mcauser/micropython-pcd8544-c/blob/5ee83f96b83236c81f11c35cd43fccf106bede11/micropython.mk

(though I bet you can find just as many examples of it *not* being included as I can of it being included, so I have to venture that Tensorflow and ulab are pretty strong candidates for setting precedent)

It's clear I'm not alone in thinking the source of truth for enabling modules should be the list of modules that are supplied to the build- since it's a user-supplied build argument. Requiring the user to enable modules in two places is a potential trap for beginners (one that, yes, I fell into when first learning to build user modules against MicroPython). I'd be inclined to add something like the line from ulab to the `micropython.mk` files for user modules to establish this as best practise.

Is there likely to be a case where a user includes a module but *doesn't* want it enabled? I guess our very liberal collection of modules might be one example. We have a main `.cmake` file that includes all of our modules: https://github.com/pimoroni/pimoroni-pico/blob/main/micropython/modules/micropython.cmake

And then we build our version of the port *just* by including this list, without needing to apply any pre-build patches to `mpconfigport.h` or maintain our own fork with a different list of enabled modules.

As an end-user if you didn't want *all the things* from our top-level `micropython.cmake` you would have to do something like `#undef MODULE_NAME_ENABLED` in `mpconfigport.h` to explicitly disable modules on a case by case basis, but the core premise of my argument is that someone building a custom MicroPython with their choice of modules:

1. Shouldn't have to edit or even care about `mpconfigport.h`
2. Shouldn't be surprised when they use only the `USER_C_MODULES` argument, but the  modules are missing from the build

I'm not particularly concerned one way or another regarding your in-house examples, since whether or not they enable automatically doesn't change things *for me* but I'm willing to make the case that the original design intent, and how the wider community have embraced it have diverged somewhat and perhaps (at the risk of making more work for myself) the example `micropython.mk` files and documentation should update to reflect this.

dpgeorge commented on ports/rp2/Makefile:
```diff
@@ -6,8 +6,17 @@ BUILD = build
+endif
+
+# If compression is enabled, trigger the build of compressed.data.h...
+OBJ_EXTRA_ORDER_DEPS += $(HEADER_BUILD)/compressed.data.h
```
I don't think this line is needed...

Gadgetoid replied:
Whoops, that crept in somewhere!

mocleiri commented:
@Gadgetoid I've reworked #6956 so that it sits on top of these commits.  

Thanks to all of the work done on this pull request the [changes](https://github.com/micropython/micropython/pull/6956/commits/14f9e2fad1424f95e01aa9f3f3f6242b19123f03) are very small now on the esp32 side. 

Special thanks for figuring out the qstr generation for cxx modules as that was what was breaking for me before but now on the latest changes it just worked.

Gadgetoid commented:
Great news @mocleiri - as always I stand on the shoulders of giants and don't think I'd have got the qstr stuff worked out without @dpgeorge's nudging (and certainly not anywhere much without Graham's additional code) but it's nonetheless a relief to know it's all working for you. Go team!

I think it's just the precedent-setting question of whether or not we follow the example of popular user modules and auto-enable modules in their `micropython.mk` and `micropython.cmake` files that remains now.

dpgeorge commented:
> I think it's just the precedent-setting question of whether or not we follow the example of popular user modules and auto-enable modules in their `micropython.mk` and `micropython.cmake` files that remains now.

Yes.  The original use for having additional config macros to enable, was for built-in modules to use this feature as well (eg the `array` module uses `MP_REGISTER_MODULE()`).  So all built-in modules are included in the build process but only those that are enabled (in mpconfigport.h) are built.

> Shouldn't be surprised when they use only the USER_C_MODULES argument, but the modules are missing from the build

I also fell into this trap!  So sometimes I cheat and just do (note the "1"):
```
MP_REGISTER_MODULE(MP_QSTR_mymodule, mp_module_mymodule, 1);
```
I'd almost prefer this pattern as the recommended default, because then there's no need to introduce any macro config value which is anyway just defaulted to enabled.

But note that you do not need to modify `mpconfigport.h`.  `CFLAGS_EXTRA` is there to support defining things on the (make) command line.  See eg `tools/ci.sh:ci_stm32_pyb_build` which uses `CFLAGS_EXTRA="-DMODULE_CEXAMPLE_ENABLED=1`.

Anyway, I guess the strongest argument is user convenience.  It is a trap that including a C user module in the build is not enough, it must also be enabled.

I think we can merge this PR now, as-is, and have a separate PR/discussion (if needed) about the default.

Gadgetoid commented:
> I think we can merge this PR now, as-is, and have a separate PR/discussion (if needed) about the default.

I would tend to agree, since it's another bike-shedding rabbit hole. (And also, of course, I'd *love* to see the back of this PR and maintaining my own fork for CI purposes!)

Should I raise an issue to raise this as a potential user pitfall?- the increased adoption that RP2 will drive (I mean, hello, it sent me falling down this rabbit hole!) will no doubt necessitate some accommodations like this.

Thanks for all your input here, and - of course - to all involved! It's been a roller-coaster.

dpgeorge commented:
Ok, merged in ccc388f157eacfea6b5c44e1b6049a2bbeb44734 through 5976ea02a50ea76d72ada40db4cda186147fb412

Thanks all involved for the work!

dpgeorge commented:
See #7086 for a follow up.

technical-publishing commented:
Hello everybody,
I have tried to build the rp2 port with some extension (which works for stm32) but the external files are not included in my build. So I was very glad to find this issue and to see that the problem is been solved now. I have updated my repository, but now I got another error even when I try to build the firmware including the cexample module. I have done this build on Windows WSL system and on a Ubuntu Linux machine but getting the same result. These are only the last few lines of the output:

```
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
Including User C Module(s) from ../../examples/usercmodule
CMake Error at /home/werner/micropython/py/usermod.cmake:42 (include):
  include could not find load file:

    ../../examples/usercmodule
Call Stack (most recent call first):
  CMakeLists.txt:30 (include)


Found User C Module(s):
ELF2UF2 will need to be built
-- Configuring incomplete, errors occurred!
See also "/home/werner/micropython/ports/rp2/build/CMakeFiles/CMakeOutput.log".
See also "/home/werner/micropython/ports/rp2/build/CMakeFiles/CMakeError.log".
make: *** [Makefile:16: all] Error 1
```
I hope this is the right place for this message, otherwise please accept my apologies. I would be very grateful for any advice.
Thanks Werner

Gadgetoid commented:
The path you supply should be something like `../../examples/usercmodule/micropython.cmake` since it was updated to require the full path to the CMake file rather than just the directory.

technical-publishing commented:
Thanks so much! Now it works... Is it consistent for all ports? I am working with STM32 and ESP32 too.

dpgeorge commented:
> I had a look into cmodules.rst but there is only still a reference to the directory, also for cmake based ports like rp2?

You are right, the docs are wrong.  This should be fixed by #7092.
